### PR TITLE
Use wheels instead of file_packager

### LINKED ISSUE #655: Binary wheels for pyodide?
There was earlier discussion about this in https://github.com/iodide-project/pyodide/issues/38 and https://github.com/iodide-project/pyodide/issues/344 but I think it could be useful to have a separate issue to discuss what binary wheels for pyodide could look like.

Python wheels are basically just zip files with the package contents, see [PEP 427](https://www.python.org/dev/peps/pep-0427/). We are currently already supporting pure python wheels.

We could also consider trying to build wheels for packages that have C extensions. I'm not exactly sure what would be necessary on the technical side to make it work, and run the wasm binary through the compiler when the modules are loaded. That would require more investigation, and it could be a blocker, but I imagine it could be made to work. 

For instance a numpy wheel could be,
```
 numpy-1.4.0-py37-pyodide015-wasm.whl
```
where the latter two parts are the binary ABI and the platform.

**Advantages**
 - wheels have already a clear spec, deal with ABI compatibility and are natural for python users.
 - I think they could be distributed through PyPi (or any other 
warehouse / PyPi index). If the platform tag is not supported now, I wouldn't likely be too difficult to support it.
 - Because it's just a zip file, it should be relatively easy to repackage a set of wheels into one or strip unused parts for a given project (without dealing with emscripten side of things at all).
 - They might make package building a bit more sustainable, because we would have to deal less with it in pyodide, and users can do it and distribute results independently.
 - One file per package to download instead of two (`.js` + `.data`).
 
**Disadvantages**
 - it puts stronger constraints on ABI compatibility, meaning that we cannot make releases every other month that breaks compatibility (or in particular update emscripten). That's certainly not ideal, and we considered it a blocker in the past, https://github.com/iodide-project/pyodide/issues/344 but we do have to start to deal with it at some point in any case probably.
 - in particular, it likely means we cannot make releases too often. For the situation with PyPy wheels https://github.com/antocuni/pypy-wheels is a relevant reference.
 - related to the previous point, it makes re-generation of a full distribution more difficult, as one has lots of distinct packages.
 - I'm not sure that a zip format is ideal for the web, but at the same time we do need some container to distributes multiple files in a python package.

Again I'm not saying that we should migrate to using wheels, but it could be interesting to consider what would be necessary to make it work.

### PR DESCRIPTION
**Pull request by hoodmane**
Use wheels. Resolves #655.

In the process, I rewrote most of `load-pyodide`, eliminating all the remaining black magic and confusing control flow.

It is worth pointing out that it is perfectly possible now to make `package.json` point at pypi, and by doing this get automatic installation into the repl for a preselected collection of pypi packages. The only difference between `loadPackage` and `micropip` is the dependency resolution. `loadPackage` is designed for "static" dependency resolution, whereas `micropip` is designed for "dynamic" dependency resolution. For this reason, it might be good to encourage use of `loadPackage`. I think ideally what we could do is make `micropip.freeze` which dumps a new `packages.json` with the currently loaded set of packages added in. Then a user can get the setup they want and freeze it so they won't have to redo dependency resolution every time their app starts.

### CODE REVIEW

**Comment by hoodmane** [comment]
In particular, I think `loadPackage` should be changed to *only* load packages in `package.json`. For loading anything else, we should use `micropip`. This will lead to a clear distinction in purposes between these two tools.

**Comment by hoodmane** [comment]
Probably we need an RFC for this of some sort.

**Comment by hoodmane** [comment]
Screenshot from 2021-12-10 08-54-24
Most of the packages are building correctly.

**Comment by hoodmane** [comment]
I don't think the `needs_rebuild` command is working correctly. It seems to rebuild stuff a lot even when it isn't needed.

**Comment by hoodmane** [comment]
Okay at least locally all 89 package builds are working.

**Comment by hoodmane** [comment]
Now all everything builds correctly, but there is some problem with test collection. All package tests get SKIPPED, while the micropip tests have a collection failure. Not sure why this happens.

**Comment by hoodmane** [comment]
@rth Could we mark url arguments to `pyodide.loadPackage` as deprecated in v0.19 in preparation for this?

**Comment by hoodmane** [comment]
Interestingly core tests only seem messed up on circleci, not on github actions.

**Comment by hoodmane** [comment]
Okay deleting the unpacked test and wheel directories fixed it.

**Comment by hoodmane** [comment]
Okay the core problems are now down to:
1. test_cpython_core ast is mysteriously failing
2. test_package_loading load_failure_retry needs to be rewritten or removed
3. node tests has `fetch` is not defined name errors 

Packages:
1. CLAPACK raises a wasm compilation error "CompileError: wasm validation error: at offset 4008999: failed to start custom section" which breaks all dependent packages
2. a few packages try to import from deleted test files.

**Comment by hoodmane** [comment]
Strangely, clapack works fine if I zip it, but it breaks if I tar it. It seems that the untar utility is buggy in some way...

**Comment by hoodmane** [comment]
Strangely, clapack works fine if I zip it, but it breaks if I tar it. It seems that the untar utility is buggy in some way...

**Comment by hoodmane** [comment]
Okay, indeed now that I am zipping CLAPACK, it works fine. Some day maybe we'll have to investigate what the heck is going wrong with untarring .so files...

**Comment by hoodmane** [comment]
Okay I think this is all set now.

File: packages/Makefile
Comment (ryanking13) [inline packages/Makefile:25]: Maybe add some comment here to tell why we use custom fork of wheel and when it will be fixed?

Comment (hoodmane) [reply packages/Makefile:25]: Right. It looks like my PR will be merged soon probably:
https://github.com/pypa/wheel/pull/427

Comment (hoodmane) [reply packages/Makefile]: Okay I can actually just use `find -name '*,*' -delete`. Somehow I never think of the simple solution until after I've implemented a complicated solution =P

**Comment by rth** [comment]
Great progress @hoodmane ! I'll try to review in the next few days.

**Comment by hoodmane** [comment]
Okay now this is great: the node test that is failing passes if you redirect standard error.

File: Makefile
Comment (rth) [inline Makefile:177]: Should we exclude __pycache__ just in case?

File: packages/mne/patches/0001-Force-mne-to-accept-old-scipy.patch
Comment (rth) [inline packages/mne/patches/0001-Force-mne-to-accept-old-scipy.patch:18]: Can probably be removed now

File: packages/statsmodels/meta.yaml
Comment (rth) [inline packages/statsmodels/meta.yaml:14]: I image those are not important files?

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:42]:
```suggestion
def _make_whlfile(*args, owner=None, group=None, **kwargs):
```
to keep it private

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:439]:
```suggestion
    shutil.rmtree(test_dir, ignore_errors=True)
```

File: src/py/pyodide/package_loader.py
Comment (rth) [inline src/py/pyodide/package_loader.py:1]: Functions need documentation type annotation, and maybe let's make this `_package_loader` if we don't intend to make it public?

File: src/tests/test_package_loading.py
Comment (rth) [inline src/tests/test_package_loading.py:7]: This should still work after the file format change to avoid breaking backward compatibility, unless there is a reason to remove it?

**Review by rth — COMMENTED** [review]
Thanks for working on this! A first round of review is below. I reviewed everything except for src/js/load-pyodide.js where the diff is quite large and maybe let's discuss the general comments before going into details there.

> This introduces significant backwards incompatibility, so we should discuss the best way to migrate. 

Is this unavoidable in this PR? Can't we have only build & .js + (.data  -> .whl) format change and keep the dependency resolution mostly as it was? About static vs dynamic dependency resolution, I agree we need to move in that direction but it's a fairly complex design decision and I think we need to spend more time iterating on what usage should look like  (with examples) and get user feedback on that. So that would be better done in a separate issue, and it would have been ideal to switch to wheels separately from that discussion.

File: packages/mne/patches/0001-Force-mne-to-accept-old-scipy.patch
Comment (hoodmane) [reply packages/mne/patches/0001-Force-mne-to-accept-old-scipy.patch:18]: =)

File: packages/statsmodels/meta.yaml
Comment (hoodmane) [reply packages/statsmodels/meta.yaml:14]: No, they are test data files. But we actually fixed this twice over:
1. They are not in stats models 0.13.1
2. wheel package v0.37.1 contains my fix for the bug https://github.com/pypa/wheel/pull/427

**Comment by ryanking13** [comment]
> Is this unavoidable in this PR? Can't we have only build & .js + (.data -> .whl) format change and keep the dependency resolution mostly as it was? About static vs dynamic dependency resolution, I agree we need to move in that direction but it's a fairly complex design decision and I think we need to spend more time iterating on what usage should look like (with examples) and get user feedback on that. So that would be better done in a separate issue, and it would have ideal to switch to wheels separately from that discussion.

I agree, this PR is trying to do too many things at once, so I'm afraid we might miss something important.

File: pyodide-build/pyodide_build/buildpkg.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildpkg.py]: Currently this is handled in buildall:
```py
        if self.shared_library:
            file_path = shutil.make_archive(
                f"{self.name}-{self.version}", "zip", self.pkgdir / "dist"
            )
            shutil.copy(file_path, outputdir)
            return
```
Maybe we should zip the file in buildpkg instead though? And of course we could use a tar file instead of a zip file if you prefer.

File: Makefile
Comment (hoodmane) [reply Makefile]: Sure, done.

File: pyodide-build/pyodide_build/buildpkg.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildpkg.py]: Okay, done.

**Comment by hoodmane** [comment]
Why is the format check failing on `_micropip.py`?

**Comment by ryanking13** [comment]
> Why is the format check failing on `_micropip.py`?

You should update your local docker image (or local black version)
See https://github.com/pyodide/pyodide/pull/2084

**Comment by hoodmane** [comment]
> You should update your local docker image 

I'm confused about this: I see `pyodide/pyodide-env:19` in `config.yml` and `PYODIDE_IMAGE_TAG="19"` in `run_docker` and I am running `pyodide/pyodide-env:19` locally. So why is the version of `black` different?

**Comment by ryanking13** [comment]
> > You should update your local docker image
> 
> I'm confused about this: I see `pyodide/pyodide-env:19` in `config.yml` and `PYODIDE_IMAGE_TAG="19"` in `run_docker` and I am running `pyodide/pyodide-env:19` locally. So why is the version of `black` different?

We updated the docker image with the same tag, you have to remove your local image (`docker rmi`) and pull it again.
Maybe we better use a different tag whenever we push a updated docker image.

File: src/tests/test_package_loading.py
Comment (hoodmane) [reply src/tests/test_package_loading.py]: Okay fixed.

**Comment by hoodmane** [comment]
@rth Okay, I think I addressed the first round of comments. At least I fixed the dependency overrides at least. The dependency overrides aren't terrible, but at some later point it would be good to consider whether to change our approach to them.

> `src/js/load-pyodide.js` where the diff is quite large

I split some of those changes into #2106, I will try to move more over there.

**Comment by hoodmane** [comment]
Okay, so about 50% of the changes to `load-pyodide.js` I pulled out into #2106 (which no longer is accurately named hmm).

**Comment by hoodmane** [comment]
Apparently setuptools comes with 700kb of windows binaries.

File: packages/distlib/meta.yaml
Comment (rth) [inline packages/distlib/meta.yaml:7]: Do we need this parameter? We could pretty much look at the extension and the file ending with `none-any` couldn't we ?

File: pyodide-build/pyodide_build/buildall.py
Comment (rth) [inline pyodide-build/pyodide_build/buildall.py:38]:
```suggestion
    install_dir: Path = Path("site")
```
?

File: pyodide-build/pyodide_build/buildall.py
Comment (rth) [inline pyodide-build/pyodide_build/buildall.py:482]: That's a bit wide. Maybe let's at least put the `pkg.name.replace('-', '_')` and the version in the glob and make a case insensitive search. Then check that there is only one result? Or alternatively validate the path with https://packaging.pypa.io/en/latest/utils.html#packaging.utils.parse_wheel_filename

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:312]: We could use it directly without subprocess (with import of wheel inside this function) https://github.com/pypa/wheel/blob/c716c87ffa528ac9a31de3ed3217058848dc2cd2/src/wheel/cli/unpack.py however it's unclear how stable their public API is. I would guess it shouldn't be too high for these two functions https://github.com/pypa/wheel/issues/262

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:409]: Nit: docstrings need a 1 line description followed by an empty line )

```suggestion
    """Package a wheel
    
    Parameters
```

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:407]: Could you add type annotations, please ?

File: src/js/load-pyodide.js
Comment (rth) [inline src/js/load-pyodide.js:428]: Very minor: Maybe better to define a variable that's going to be passed to downloadPackage, to avoid repeating this line twice?

File: src/py/pyodide/_package_loader.py
Comment (rth) [inline src/py/pyodide/_package_loader.py:23]:
```suggestion
    """Used to install a package either into sitepackages or into the standard library.
    
    This is a helper method called from ``loadPackage``.
```
Assuming the first line is not too long.

File: src/py/pyodide/_package_loader.py
Comment (rth) [inline src/py/pyodide/_package_loader.py:20]: filename as Path should also work I think?

File: src/py/pyodide/_package_loader.py
Comment (rth) [inline src/py/pyodide/_package_loader.py:54]:
```suggestion
    """List out the paths to .so files in a zip or tar archive.
```

File: src/py/pyodide/_package_loader.py
Comment (rth) [inline src/py/pyodide/_package_loader.py:84]: Can we add a test for this (running in system python not Pyodide)?

**Review by rth — COMMENTED** [review]
Thanks a lot for doing this @hoodmane ! It looks great and I'm very much looking forward to using it.

A few comments below but overall it looks good to me.
The file sizes seem to be much smaller,
```
-rw-r--r-- 1 root root 7.1M Jan 19 11:48 numpy.data
[...]
-rw-r--r-- 1 root root 3.4M Jan 14 00:52 numpy-1.21.4-cp39-cp39-emscripten_wasm32.whl
```
we will see what's the impact after brotli compression but I though that .data was initially gzip compressed as well? So it looks like the gzip compression is significantly higher?
In some later time, we might want to experiment whether it's better to create wheels with the default zip compression or with 0 compression and let the browser compression be used instead.

About load times, our benchmarks are not too reliable but, we have

```
 -  load numpy  (main)
    native: nan  firefox: 0.938628  chrome: 3.942906
 -  load numpy  (this PR)
    native: nan  firefox: 0.837011  chrome: 3.967529 
```
so overall it looks comparable.

File: pyodide-build/pyodide_build/buildall.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:38]: Well this is a choice between "site" ==> `SITE_PACKAGES` and "lib" ==> `STD_LIB`. I look up the proper locations for `SITE_PACKAGES` from `site.getsitepackages()` and for `STD_LIB` from `sysconfig.get_path("stdlib")` at the top of the file. Not sure if using a string for the setting is standard or what.

File: pyodide-build/pyodide_build/buildall.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:482]: How do I do case insensitive search?

Comment (rth) [reply pyodide-build/pyodide_build/buildall.py:482]: Maybe https://stackoverflow.com/a/8151371/1791279 but yeah it looks not obvious. Just do something to avoid the situation where multiple wheels end up being in that folder and we pick the wrong one.

File: src/js/load-pyodide.js
Comment (hoodmane) [reply src/js/load-pyodide.js]: Yeah this is a bit sloppy. I rearranged the logic a bit, I think it's better now.

File: pyodide-build/pyodide_build/buildall.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:482]: How about for now we add an assert that it only contains one `.whl` file? The options should either be a `.whl`, a `.whl` and a `.tar` with unvendored tests, or a `.tar` with a shared lib.

Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py:482]: I guess maybe if you update the version of the package or something, we will end up with two wheel files in the `dist` directory. Probably we'd want to pick the right wheel in that case rather than crashing. Or we could empty the `dist` directory before putting the new stuff into it...

File: src/py/pyodide/_package_loader.py
Comment (hoodmane) [reply src/py/pyodide/_package_loader.py:20]: Yeah it would work, though we are calling this from Javascript so in practice it's gonna be a string.

File: src/py/pyodide/_package_loader.py
Comment (hoodmane) [reply src/py/pyodide/_package_loader.py:85]: Sure. What's the best way to make the test files? I guess we could generate them with Python code in the test, would probably be easier just to include a wheel and a tar file in some `test_data` directory though.

File: packages/distlib/meta.yaml
Comment (hoodmane) [reply packages/distlib/meta.yaml:7]: Yeah, that's definitely possible, perhaps that'd be better. After we've merged this PR we should change all the unpatched pure Python packages to use this method.

File: src/py/pyodide/_package_loader.py
Comment (hoodmane) [reply src/py/pyodide/_package_loader.py:85]: Even with patches, we could potentially download wheel, unpack wheel, apply patch, repack wheel. Then again, if we do this to avoid handling the alternate build backends used by `typing_extensions` and `distlib`, maybe we'd be better off actually figuring out how to handle alternate build backends. It doesn't look that hard.

File: packages/distlib/meta.yaml
Comment (hoodmane) [reply packages/distlib/meta.yaml:7]: Even with patches, we could potentially download wheel, unpack wheel, apply patch, repack wheel. Then again, if we do this to avoid handling the alternate build backends used by `typing_extensions` and `distlib`, maybe we'd be better off actually figuring out how to handle alternate build backends.

Comment (hoodmane) [reply packages/distlib/meta.yaml:7]: Also, would be good to update `mkpkg` to prefer to take pure Python wheels over sdists. Maybe with a flag to indicate that we actually want an sdist. For updates, can default to keeping the same type.

File: pyodide-build/pyodide_build/buildall.py
Comment (hoodmane) [reply pyodide-build/pyodide_build/buildall.py]: Okay, now I explicitly assert that there is only one wheel in the dist directory. I think this makes it less error prone.

File: src/py/pyodide/_package_loader.py
Comment (rth) [reply src/py/pyodide/_package_loader.py:85]: Generating them at test time in some tmp folder would be better IMO: easier to both know what's inside and change something if needed.

**Comment by hoodmane** [comment]
@rth @ryanking13 Okay, tests pass again. Any further comments? How do you feel about merging this?

**Review by rth — APPROVED** [review]
The "src/tests/data/test-scipy.{whl,tar.bz2}" files could be removed if they are not used?

Otherwise LGTM. Thanks!

File: conftest.py
Comment (ryanking13) [inline conftest.py:79]:
```suggestion
    This function lists the names of the .whl files in the build/ directory.
```

File: packages/test_packages_common.py
Comment (ryanking13) [inline packages/test_packages_common.py:33]: Would it be hard to avoid having two same functions `built_packages` in both here and in `conftest.py`?
I'm not sure which would be the best place to put this function (maybe somewhere inside `pyodide-build`?)

File: packages/distlib/meta.yaml
Comment (ryanking13) [inline packages/distlib/meta.yaml:3]: It needs a changelog for version update

**Review by ryanking13 — APPROVED** [review]
Thanks for your huge effort! I have some minor comments, otherwise LGTM. And please update the changelog for this.

**Comment by ryanking13** [comment]
We should also update the documentation which refers `.data`, but I think we could do that in a separate PR since this PR is already big enough.

File: packages/test_packages_common.py
Comment (hoodmane) [reply packages/test_packages_common.py:33]: We can just import it directly from conftest actually.

**Comment by rth** [comment]
I really like that one can now just call `less` to list files in a wheel or .tar file, and `zcat` to get wheels contents. It makes debugging so much easier. Thanks again @hoodmane !

**Comment by henryiii** [comment]
Is there a documented process somewhere for a user to build wheels? If I have a simple C/C++ package, could I produce one of these wheels now? (And ideally, the `emscripten_wasm32` tag should be allowed in PyPI, is that being pursued?)

**Comment by rth** [comment]
> Is there a documented process somewhere for a user to build wheels? 

For pure Python packages, it's just regular universal Python wheels.
For packages with C extensions, you can run inside the pyodide Docker image,
```
PYODIDE_PACKAGES="<package-name>" make
```
see the dev docs https://pyodide.org/en/latest/development/building-from-sources.html and https://pyodide.org/en/latest/development/new-packages.html.  We are trying to migrate to a better build setup & CLI. `latest` vs `stable` in the URL to get the dev docs, since this feature is not yet included in the release.
This assumes though that you have a `meta.yaml` package defining the package and `meta.yaml` for its dependencies.

> And ideally, the emscripten_wasm32 tag should be allowed in PyPI, is that being pursued?

Yes, see https://github.com/pypa/warehouse/issues/10416 . Though it's likely a multi-year project. The feedback we got from core devs is that first CPython needs to officially support WASM (which hopefully would happen for the next release), then there need to be a PEP and enough traction/usage for Python with WASM from the community.  So the short term goal is just to deploy wheels to some other wheel services (or keep our custom deployment). A bit similar to what is done for https://github.com/antocuni/pypy-wheels

Having the ability to build wheels for WASM as part of cibuildwheel would be indeed ideal. I'll comment more on that in https://github.com/pypa/cibuildwheel/issues/1002

**Comment by henryiii** [comment]
I'd be quite fine with that - calling this still exploratory, keeping the ABI from gaining requirements quite yet. I think it still could be "standalone" via cibuildwheel & probably docker.

**Pull request merged**
